### PR TITLE
Preserve unusual unicode whitespace

### PR DESCRIPTION
**Pull request by karl**
This PR preserves unusual unicode whitespace characters.

Currently any whitespace that matches a unicode whitespace character gets converted to a normal space. This means we lose whitespace characters with special meanings (such as non-breaking spaces and zero width spaces).

In this PR we no longer treat anything other than newline, tab, and space characters as whitespace. This allows us to preserve these unusual unicode whitespace characters as is.

With this change I believe we no longer need to preserve multiple space characters when formatting JSX, because they can no longer contain unusual whitespace we know that they will collapse down to a single space when rendering so it is safe to do that in Prettier.

### CODE REVIEW

**Comment by karl** [comment]
This is probably not ready to merge yet, I just wanted to put it here to gather feedback.

**Comment by vjeux** [comment]
This is way out of my comfort zone, who do you think should review this? If no one, I'd be happy to just merge it if you believe this is the right thing to do :)

**Comment by lydell** [comment]
This is only about JSX, right?

**Comment by karl** [comment]
Yep, just for text in JSX!

I haven't tested normal JSX strings, but I assume they keep unusual Unicode spaces intact.

**Comment by karl** [comment]
@vjeux This is also a little out of my comfort zone 😀

I'm pretty sure it is the right thing to do, I just want to test that my assumptions about JSX are true by testing the behaviour of a range of white space characters. Once I've done that then this should be good to merge!

**Comment by lydell** [comment]
What is a "normal jSX string"?

Where did you find the `[ \f\n\r\t\v]` set of characters?

**Comment by lydell** [comment]
This test file suggests that only sequences of `[ \t\r\n]` are usually rendered as a single space (not `[\f\v]`).

```html
<!DOCTYPE html>
<meta charset="utf8">
<title>space test</title>

<p>space b</p>
<p>\fb</p>
<p>\n
b</p>
<p>\r
b</p>
<p>\t	b</p>
<p>\vb</p>
<p>\u00a0 b</p>

<hr>

<p>space   b</p>
<p>\f  b</p>
<p>\n 
 b</p>
<p>\r 
 b</p>
<p>\t 	 b</p>
<p>\v  b</p>

<hr>

<p>\u00a0   b</p>
```

**Comment by karl** [comment]
Thanks, that is the test case I was about to whip up!

I got the character range by taking all the white space chars that match the `\s` regex that are in the standard ASCII range.

I suspected I didn't have the correct range, so wanted to test before requesting a review of this PR. I'll update the PR based on your tests!

**Comment by karl** [comment]
"normal JSX strings" is a typo (adding these comments from my phone) it was meant to say "normal JS strings"!

**Comment by lydell** [comment]
I’ve read up a bit on whitespace handling in JSX.

Apparently, it is not part of the JSX spec: https://github.com/facebook/jsx/issues/19#issuecomment-57187084. Unlike JS whitespace, which is discarded while parsing, JSX whitespace is preserved in the AST nodes containing JSX text. For example, try this in the REPL and you’ll see it:

```js
require("babylon").parse("<div>\n  three spaces:   end\n</div>")
```

It is instead up to the tool, such as Babel, that transforms JSX AST nodes to JS AST nodes (function calls) to do stuff with the whitespace.

Playing around in the Babel REPL confirms that [Babel does not treat `\f` and `\v` as whitespace](http://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=es2015%2Creact%2Cstage-2%2Cstage-3&targets=&browsers=&builtIns=false&debug=false&code=%3Cdiv%3E%0A%3Cp%3E%0A%20%20%20%20%20%5Cx20%20stripped%0A%3C%2Fp%3E%0A%3Cp%3E%0A%20%20%20%20%20%20%5Ct%20stripped%0A%3C%2Fp%3E%0A%3Cp%3E%0A%20%20%0C%20%20%5Cf%20not%20stripped%0A%3C%2Fp%3E%0A%3Cp%3E%0A%20%20%0B%20%20%5Cv%20not%20stripped%0A%3C%2Fp%3E%0A%3C%2Fdiv%3E&experimental=false&loose=false&spec=false&playground=true). We should probably read Babel’s JSX transformer source code to be 100% sure.

Also, [Babel does not collapse spaces between words](http://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=es2015%2Creact%2Cstage-2%2Cstage-3&targets=&browsers=&builtIns=false&debug=false&code=%3Cdiv%3Ea%20%20%20b%3C%2Fdiv%3E&experimental=false&loose=false&spec=false&playground=true). Prettier 1.3.1 respects that, but latest master collapses it into 1 space. Perhaps I should open a separate issue about this?

So in theory we can’t change JSX formatting at all without changing the meaning the of the program. But it would be super sad to give up all our super nice JSX formatting! After all, JSX is most commonly used with Babel anyway.

**Comment by karl** [comment]
Interesting stuff,! I didn't know that the JSX spec treated whitespace as significant!

I'm not sure what that means for Prettier or this PR. Are we happy for Prettier to keep treating JSX whitespace as something that can be modified?

I've updated the PR to only count `[ \t\r\n]` as collapsable whitespace, and added a separate set of tests for the whitespace behaviour. If we are happy to continue to modify JSX whitespace then this PR is ready for review!

**Comment by lydell** [comment]
1. I’ve opened a separate issue for the spaces between words thing.
2. This PR is a big improvement to the current state.
3. I think JSX formatting is so useful that it doesn’t matter that it is theoretically changing the meaning of the program. Until somebody reports problems because of it, we should definitely keep it and treat it as a theoretical issue. That’s my opinion at least :)

**Comment by karl** [comment]
This PR currently removes duplicate spaces between JSX elements. Based on @lydell's findings it looks like we will probably need to keep them.

I'll update the PR when I have a moment, but until then consider this unsafe to merge!

**Comment by karl** [comment]
I've updated this PR to keep the existing behaviour of maintaining multiple whitespaces between tags.

It is now ready for review and merge 😀

**Comment by vjeux** [comment]
@karl awesome, thanks a ton for powering through this! @lydell could you review and merge it if it's good? Thanks!

@karl I know it's a lot to ask but would you be able to look into https://github.com/prettier/prettier/issues/1581? This is blocking for the next release at fb and I have no idea about that part of the code :p

**Comment by karl** [comment]
Happy to take a look at #1581. Might be a few days until I have the time, but I don't think it should be too difficult to fix.

**Comment by lydell** [comment]
Heh. So #1581 looks like a case where "we can’t change JSX formatting at all without changing the meaning the of the program", but it looks fixable without being so drastic.

File: tests/jsx-whitespace/test.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Should collapse
```
Comment (lydell) [inline tests/jsx-whitespace/test.js]: Same here.

File: tests/jsx-whitespace/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -0,0 +1,55 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`test.js 1`] = `
+// Should collapse
```
Comment (lydell) [inline tests/jsx-whitespace/__snapshots__/jsfmt.spec.js.snap]: These shouldn't collapse. If it's out of scope for this PR, we could just change the comment.

http://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=es2015%2Creact%2Cstage-2%2Cstage-3&targets=&browsers=&builtIns=false&debug=false&code=spaces%20%3D%20%3Cdiv%3E%5D%20%20%20%5B%3C%2Fdiv%3E&experimental=false&loose=false&spec=false&playground=true

File: tests/jsx_escape/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -46,7 +46,7 @@ raw_amp = <span>foo & bar</span>
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 many_nbsp = <div>&nbsp; &nbsp; </div>;
 single_nbsp = <div>&nbsp;</div>;
-many_raw_nbsp = <div>   </div>;
+many_raw_nbsp = <div>   </div>;
```
Comment (lydell) [inline tests/jsx_escape/__snapshots__/jsfmt.spec.js.snap]: These have turned from regular spaces back to non-breaking spaces, right? That's great!

File: tests/jsx-whitespace/__snapshots__/jsfmt.spec.js.snap (hunk 2 above)
Comment (karl) [inline tests/jsx-whitespace/__snapshots__/jsfmt.spec.js.snap]: Good catch!

Comment (karl) [inline tests/jsx-whitespace/__snapshots__/jsfmt.spec.js.snap]: I've tweaked the comments to remove any mention of collapsing.

**Comment by vjeux** [comment]
Yay, awesome work on this :)

**Pull request merged**